core/src/test/java/io/grpc/internal/ManagedChannelImplTest.java

```diff
@@ -4476,6 +4476,88 @@ public String getDefaultScheme() {
+              EquivalentAddressGroup(socketAddress))).setError(resolutionError).build();
+
+      channelBuilder.nameResolverFactory(nameResolverFactory);
+
+      Map<String, Object> defaultServiceConfig =
+              parseConfig("{\"methodConfig\":[{"
+                      + "\"name\":[{\"service\":\"SimpleService1\"}],"
+                      + "\"waitForReady\":true}]}");
+
+      channelBuilder.defaultServiceConfig(defaultServiceConfig);
+      createChannel(true);
+
+      int prevSize = getStats(channel).channelTrace.events.size();
+
+      assertThat(getStats(channel).channelTrace.events.get(prevSize - 1))
+              .isEqualTo(new ChannelTrace.Event.Builder()
+              .setDescription("Initial Name Resolution error, using default service config")
+              .setSeverity(ChannelTrace.Event.Severity.CT_ERROR)
+              .setTimestampNanos(timer.getTicker().read())
+              .build());
+
+      nameResolverFactory.resolvers.get(0).listener.onError(resolutionError);
+
+      prevSize = getStats(channel).channelTrace.events.size();
+
+      assertThat(getStats(channel).channelTrace.events).hasSize(prevSize);
+      assertThat(getStats(channel).channelTrace.events.get(prevSize - 1))
+              .isEqualTo(new ChannelTrace.Event.Builder()
+              .setDescription("Initial Name Resolution error, using default service config")
+              .setSeverity(ChannelTrace.Event.Severity.CT_ERROR)
+              .setTimestampNanos(timer.getTicker().read())
+              .build());
+    } finally {
+      LoadBalancerRegistry.getDefaultRegistry().deregister(mockLoadBalancerProvider);
+    }
+  }
+
+  @Test
+  public void testTraceEventsOnResultsCallFollowedByOnErrorCallWithAppliedServiceConfig()
+          throws Exception {
+    timer.forwardNanos(1234);
+    channelBuilder.maxTraceEvents(10);
+    List<EquivalentAddressGroup> servers = new ArrayList<>();
+    servers.add(new EquivalentAddressGroup(socketAddress));
+    FakeNameResolverFactory nameResolverFactory =
+            new FakeNameResolverFactory.Builder(expectedUri).setServers(servers).build();
+
+    channelBuilder.nameResolverFactory(nameResolverFactory);
+    Map<String, Object> defaultServiceConfig =
+            parseConfig("{\"methodConfig\":[{"
+                    + "\"name\":[{\"service\":\"SimpleService1\"}],"
+                    + "\"waitForReady\":true}]}");
+
+    channelBuilder.defaultServiceConfig(defaultServiceConfig);
+    createChannel();
+    int prevSize = getStats(channel).channelTrace.events.size();
+
+    assertThat(getStats(channel).channelTrace.events).hasSize(prevSize);
+
+    prevSize = getStats(channel).channelTrace.events.size();
+    Status resolutionError = Status.UNAVAILABLE
+            .withDescription("Initial Name Resolution error, using default service config");
+
+    nameResolverFactory.resolvers.get(0).listener.onError(resolutionError);
+
+    assertThat(getStats(channel).channelTrace.events).hasSize(prevSize + 1);
```

Comment:
Add a comment above this assert that since initial service config is already applied, it is not reapplied using the default service config here.